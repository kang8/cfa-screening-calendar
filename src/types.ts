export type Movie = {
  name: string
  englishName: string
  year: number
  director: string
  cinima: string
  playTime: string
  endTime: string
}

export type CinemaInfo = '小西天艺术影院' | '百子湾艺术影院' | '江南分馆影院'

export type MovieHall = '1号厅' | '2号厅'

export type movieActor = {
  position: '导演' | '编剧' | '演员'
  realName: string
}

export type Body = {
  data: {
    records: [{
      movieInfo: {
        movieName: string
        englishName: string
        movieTime: string
        movieMinute: number
      }
      cinemaInfo: CinemaInfo
      movieHall: MovieHall
      playTime: string
    }]
  }
  total: string
  size: string
}
