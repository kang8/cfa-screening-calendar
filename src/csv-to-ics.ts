import { createEvents, DateArray, EventAttributes } from 'npm:ics@^3.5.0'
import { encodeHex } from 'https://deno.land/std@0.202.0/encoding/hex.ts'
import { parse as csvParse } from 'https://deno.land/std@0.184.0/csv/parse.ts'
import { Movie } from './types.ts'
import { sortByPlayTime } from './json-to-csv.ts'

export const csvDir = './assets/csv'

function parseDate(date: Date): DateArray {
  return [
    date.getFullYear(),
    date.getMonth() + 1,
    date.getDate(),
    date.getHours(),
    date.getMinutes(),
  ]
}

export function parseTitle(movie: Movie): string {
  if (movie.englishName.trim() === '') {
    return `${movie.cinima}|${movie.name}`
  }

  return `${movie.cinima}|${movie.name}(${movie.englishName})`
}

export async function csvToIcs(filenames: string[]) {
  for await (const filename of filenames) {
    const movies: Movie[] = []
    const content = await Deno.readTextFile(`${csvDir}/${filename}.csv`)

    movies.push(
      ...csvParse(content, { skipFirstRow: true }) as unknown as Movie[],
    )
    sortByPlayTime(movies)

    const events: EventAttributes[] = await Promise.all(
      movies.map((movie) =>
        crypto.subtle.digest(
          'SHA-1',
          new TextEncoder().encode(parseTitle(movie)),
        ).then((hashBuffer) => {
          const uid = encodeHex(hashBuffer).slice(0, 6) + '_' +
            new Date(movie.playTime).getTime().toString().slice(0, 8)

          return {
            calName: '电影资料馆',
            uid: uid,
            timestamp: '20111111T000000Z',
            start: parseDate(new Date(movie.playTime)),
            end: parseDate(new Date(movie.endTime)),
            title: parseTitle(movie),
            description: 'TBD',
          }
        })
      ),
    )

    createEvents(events, async (error: Error | undefined, value: string) => {
      if (error) throw error

      await Deno.writeTextFile(`./assets/ics/${filename}.ics`, value)
    })
  }
}
