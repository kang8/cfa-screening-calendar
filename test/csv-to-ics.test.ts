import { parse as csvParse } from 'https://deno.land/std@0.184.0/csv/parse.ts'
import { Movie } from '../src/types.ts'
import { parseTitle } from '../src/csv-to-ics.ts'
import { assertSnapshot } from 'https://deno.land/std@0.185.0/testing/snapshot.ts'

Deno.test('CSV parse', async (t) => {
  const content = await Deno.readTextFile('test/fixtures/test.csv')
  const movies: Movie[] = []

  movies.push(...csvParse(content, {
    skipFirstRow: true,
  }) as unknown as Movie[])

  await assertSnapshot(t, movies)
})

Deno.test('Parse title', async (t) => {
  const movies: Movie[] = [{
    cinima: '小西天',
    endTime: '2023-05-07 20:40:00',
    englishName: '',
    name: '弗兰兹',
    playTime: '2023-05-07 19:00:00',
    year: 2016,
    director: '弗朗索瓦·欧容',
  }, {
    cinima: '百子湾',
    endTime: '2023-05-07 18:09:00',
    englishName: 'Black Cat, White Cat',
    name: '黑猫白猫',
    playTime: '2023-05-07 16:00:00',
    year: 1998,
    director: '埃米尔·库斯图里卡 Emir Kusturiča',
  }]

  const titles = movies.map(parseTitle)

  await assertSnapshot(t, titles)
})
