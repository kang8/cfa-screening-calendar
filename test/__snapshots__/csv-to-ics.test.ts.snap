export const snapshot = {};

snapshot[`CSV parse 1`] = `
[
  {
    cinima: "小西天",
    endTime: "2023-05-20 12:57:00",
    englishName: "Love Letter",
    name: "情书",
    playTime: "2023-05-20 11:00:00",
    year: "1995",
  },
  {
    cinima: "百子湾",
    endTime: "2023-05-20 15:23:00",
    englishName: "",
    name: "弗兰兹",
    playTime: "2023-05-20 13:30:00",
    year: "2016",
  },
]
`;

snapshot[`Parse title 1`] = `
[
  "小西天|弗兰兹",
  "百子湾|黑猫白猫(Black Cat, White Cat)",
]
`;
